debian/gbvdoc.1
.\" Automatically generated by Pandoc 2.0
.\"
.TH "GBVDOC" "1" "" "MANUAL" ""
.hy
.SH NAME
.PP
gbvdoc \- GBV Documents
.SH SYNOPSIS
.PP
The application is automatically started as service, listening on port
6012.
.IP
.nf
\f[C]
sudo\ service\ gbvdoc\ {status|start|stop|restart}
\f[]
.fi
.SH INSTALLATION
.PP
The software is released as Debian package for Ubuntu 14.04 LTS.
Other Debian based distributions \f[I]might\f[] work too.
Releases can be found at <https://github.com/gbv/gbvdoc/releases>
.PP
To install required dependencies either use a package manager such as
\f[C]gdebi\f[], manually install dependencies (inspectable via
\f[C]dpkg\ \-I\ gbvdoc_*.deb\f[]):
.IP
.nf
\f[C]
sudo\ dpkg\ \-i\ ...\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ #\ install\ dependencies
sudo\ dpkg\ \-i\ gbvdoc_X.Y.Z_amd64.deb\ \ \ \ #\ change\ X.Y.Z
\f[]
.fi
.PP
After installation the service is available at localhost on port 6012.
Better put the service behind a reverse proxy to enable SSL and nice
URLs! In the following a reverse proxy mapping
<http://uri.gbv.de/document/> to <http://localhost:6012/> is assumed.
.SH USAGE
.PP
GBV Documents provides an interface with links to APIs for access of
metadata from databases used and/or provided by VZG.
.SH ADMINISTRATION
.SS Configuration
.PP
Config file \f[C]/etc/default/gbvdoc\f[] only contains basic server
configuration in form of simple key\-values pairs:
.IP \[bu] 2
\f[C]PORT\f[] \- port number (required, 6012 by default)
.IP \[bu] 2
\f[C]WORKERS\f[] \- number of parallel connections (required, 5 by
default).
.SS Logging
.PP
Log files are located at \f[C]/var/log/gbvdoc/\f[]:
.IP \[bu] 2
\f[C]error.log\f[]
.IP \[bu] 2
\f[C]access.log\f[]
.SH CHANGES
.PP
See \f[C]debian/changelog\f[].
.SH SEE ALSO
.PP
The source code of gbvdoc is managed in a public git repository at
<https://github.com/gbv/gbvdoc>.
Please report bugs and feature request at
<https://github.com/gbv/gbvdoc/issues>!
.PP
The Changelog is located in file \f[C]debian/changelog\f[].
.PP
Development guidelines are given in file \f[C]CONTRIBUTING.md\f[].


debian/changelog
gbvdoc (0.3.2) unstable; urgency=medium

  * Add parameter format=redirect

 -- Jakob Voß <jakob.voss@gbv.de>  Mon, 30 Apr 2019 12:05:00 +0100

gbvdoc (0.3.2) unstable; urgency=medium

  * Add parameter format=redirect

 -- Jakob Voß <jakob.voss@gbv.de>  Wed, 30 Jan 2018 12:30:00 +0100

gbvdoc (0.3.1) unstable; urgency=medium

  * Tidy up HTML interface and description 

 -- Jakob Voß <jakob.voss@gbv.de>  Fri, 18 Dec 2015 13:01:56 +0100

gbvdoc (0.3.0) stable; urgency=low

  * migrate to easier Debian packaging

 -- Jakob Voß <jakob.voss@gbv.de>  Tue, 17 Nov 2015 10:30:00 +0100


lib/App/GBVDoc.pm
package App::GBVDoc;
use v5.14;

our $VERSION="0.3.2";

#ABSTRACT: GBV Linked Data Server http://uri.gbv.de/document/

use Log::Contextual qw(:log);
use RDF::Lazy qw(0.061);
use Plack::Request;
use Plack::Response;

use RDF::Trine qw(iri);
use RDF::Trine::Model;
use RDF::Trine::Parser;

use RDF::NS;
our $NS = RDF::NS->new('20111102');

use RDF::Flow::Cascade;
use GBV::RDF::Source::Item;
use GBV::RDF::Source::Document;
use CHI;

use Plack::Builder;

use parent 'Plack::Component';
use Plack::Util::Accessor qw(source base formats);

use RDF::NS;
our $NS = RDF::NS->new('20111102');

use Plack::Middleware::TemplateToolkit;
use Plack::Middleware::RDF::Flow qw(0.170);
use Try::Tiny;
use CHI;

our %FORMATS = (
    nt   => 'ntriples',
    rdf  => 'rdfxml',
    xml  => 'rdfxml',
    ttl  => 'turtle',
    json => 'rdfjson',
);

sub core {
    my ( $self, $app, $env ) = @_;

    if ( ( $env->{'HTTP_X_FORWARDED_HOST'} || '' ) eq 'uri.gbv.de' ) {
        my $base = $self->base;
        $base =~ s{^http://uri.gbv.de/|/$}{}g;
        $env->{'SCRIPT_NAME'} = "/$base/" . ( $env->{'SCRIPT_NAME'} | '' );
        $env->{'REQUEST_URI'} = "/$base" . $env->{'REQUEST_URI'};
    }

    my $uri = $env->{'rdflow.uri'};

    $env->{'tt.vars'} = {} unless $env->{'tt.vars'};
    $env->{'tt.vars'}{'version'} = $VERSION;
    $env->{'tt.vars'}{'uri'}     = $uri;
    $env->{'tt.vars'}{'formats'} = [ @{ $self->formats } ];

    my $uri = $env->{'rdflow.uri'};
    my $rdf = $env->{'rdflow.data'};
    my $req = Plack::Request->new($env);

    if ( $rdf and $rdf->size ) {
        delete $NS->{uri};    # TODO: Bug in RDF::Lazy
        my $lazy = RDF::Lazy->new( $rdf, namespaces => $NS );
        $env->{'tt.vars'}->{uri} = $lazy->resource($uri);

        if ( $uri eq $self->base ) {

            # main page
        }
        else {
            $env->{'tt.path'} = '/document.html';
        }
    }

    $env->{'tt.vars'}->{error}     = $env->{'rdflow.error'};
    $env->{'tt.vars'}->{timestamp} = $env->{'rdflow.timestamp'};
    $env->{'tt.vars'}->{cached}    = 1 if $env->{'rdflow.cached'};

    $app->($env);
}

sub prepare_app {
    my $self = shift;
    return if $self->{app};

    delete $NS->{uri};
    $self->base('http://uri.gbv.de/document/');

    $self->formats( [qw(ttl json xml)] )
      unless $self->formats;

    # init
    my $source = RDF::Flow::Cascade->new(
        GBV::RDF::Source::Item->new,
        GBV::RDF::Source::Document->new,
        name => "item-or-document"
    );
    $source =
      RDF::Flow::Cached->new( $source,
        CHI->new( driver => 'Memory', global => 1, expires_in => '1 hour' ),
      );
    $self->source($source);

    $self->{app} = builder {
        enable 'Static',
          root => 'public',
          path => qr{\.(css|png|gif|js|ico)$};

        # cache everything else for 10 seconds. TODO: set cache time
        enable 'Cached',
          cache => CHI->new(
            driver     => 'Memory',
            global     => 1,
            expires_in => '10 seconds'
          );

        enable 'JSONP';

        enable sub {
            my $app = shift;
            sub {
                my $env    = shift;
                my $req    = Plack::Request->new($env);
                my $format = $req->parameters->{format};
                my $res    = $app->($env);
                if ( $format eq 'redirect' ) {
                    my $rdf = $env->{'rdflow.data'};
                    my $uri = $env->{'rdflow.uri'};
                    my $page =
                      $rdf->get_statements( iri($uri),
                        iri("http://xmlns.com/foaf/0.1/page"), undef )->next;
                    if ($page) {
                        $res = Plack::Response->new;
                        $res->redirect( $page->object->uri_value );
                        return $res->finalize;
                    }
                }
                $res;
              }
        };

        enable 'RDF::Flow',
          base         => $self->base,
          empty_base   => 1,
          source       => $self->source,
          pass_through => 1,
          formats      => \%FORMATS;

        # core driver
        enable sub {
            my $app = shift;
            sub {
                my $env = shift;
                $self->core( $app, $env );
            },;
        };

        Plack::Middleware::TemplateToolkit->new(
            INCLUDE_PATH => 'public',
            RELATIVE     => 1,          # ??
            INTERPOLATE  => 1,
            pass_through => 0,

            #            timer => $is_devel,
            request_vars => [qw(base)],
            404          => '404.html',
            500          => '500.html'
        );
    };
}

sub call {
    my ( $self, $env ) = @_;
    $self->{app}->($env);
}

1;
